balancer-js/src/modules/pools/pool-types/concerns/composableStable/join.concern.ts

```diff
@@ -109,16 +144,162 @@ export class ComposableStablePoolJoin implements JoinConcern {
+    }
+    return {
+      sortedAmountsIn,
+      scalingFactorsWithoutBpt,
+      upScaledBalancesWithoutBpt,
+      parsedAmp,
+      parsedTotalShares,
+      parsedSwapFee,
+      bptIndex,
+      parsedTokens,
+    };
+  }
+
+  doJoin(sortedValues: Params): JoinPoolAttributes {
+    const {
+      sortedAmountsIn,
+      scalingFactorsWithoutBpt,
+      upScaledBalancesWithoutBpt,
+      parsedAmp,
+      parsedTotalShares,
+      parsedSwapFee,
+      bptIndex,
+      parsedTokens,
+      slippage,
+      poolId,
+      joiner,
       value,
-      minBPTOut,
+    } = sortedValues;
+    // BPT out doesn't need downscaled or priceRate
+    const expectedBPTOut = this.doMaths(
+      sortedAmountsIn,
+      scalingFactorsWithoutBpt,
+      upScaledBalancesWithoutBpt,
+      parsedAmp,
+      parsedTotalShares,
+      parsedSwapFee
+    );
+
+    const userData = this.encodeUserData(
+      expectedBPTOut,
+      slippage,
+      sortedAmountsIn
+    );
+
+    const data = this.encodeFunctionData(
+      poolId,
+      joiner,
+      joiner,
+      parsedTokens,
+      userData.userData,
+      insert(sortedAmountsIn, bptIndex, '0') // Adds value for BPT
+    );
+
+    return {
+      ...data,
+      to: balancerVault,
+      value,
+      minBPTOut: userData.minBPTOut,
     };
-  };
+  }
+
+  // filter native asset (e.g. ETH) amounts
+  getEthValue(tokens: string[], amounts: string[]): BigNumber | undefined {
+    const values = amounts.filter((amount, i) => tokens[i] === AddressZero);
+    return values[0] ? BigNumber.from(values[0]) : undefined;
+  }
+
+  doMaths(
+    amountsIn: string[],
+    scalingFactorsWithoutBpt: bigint[],
+    upScaledBalancesWithoutBpt: string[],
+    upscaledAmp: string,
+    upscaledTotalShares: string,
+    upscaledSwapFee: string
+  ): bigint {
+    /*
+      Maths should use: 
+      - upscaled amounts, e.g. 1USDC = 1e18
+      - rates
+    */
+    const upScaledAmountsIn = _upscaleArray(
+      amountsIn.map(BigInt),
+      scalingFactorsWithoutBpt.map(BigInt)
+    );
+    const expectedBPTOut = StableMathBigInt._calcBptOutGivenExactTokensIn(
+      BigInt(upscaledAmp),
+      upScaledBalancesWithoutBpt.map(BigInt), // Should not have BPT
+      upScaledAmountsIn, // Should not have BPT
+      BigInt(upscaledTotalShares),
+      BigInt(upscaledSwapFee)
+    );
+    // BPT out doesn't need downscaled or priceRate
+    return expectedBPTOut;
+  }
+
+  sortV2(tokensIn: string[], amountsIn: string[], pool: Pool): SortedValues {
+    // This will keep ordering as read from Pool
+    const {
+      parsedTokens,
+      parsedTokensWithoutBpt,
+      parsedAmp,
+      parsedSwapFee,
+      parsedTotalShares,
+      scalingFactorsWithoutBpt,
+      upScaledBalancesWithoutBpt,
+    } = parsePoolInfo(pool);
+    if (!parsedAmp) {
+      throw new BalancerError(BalancerErrorCode.MISSING_AMP);
+    }
+    // Reorder amountsIn to match pool token order
+    const [, sortedAmountsIn] = reorderArrays(
```

Comment:
should be [sortedAmountsIn] instead of [, sortedAmountsIn]

Reply:
Thanks! `reorderArrays` comment was a bit confusing as output said it was returning original too. Updated that too.